### DISCUSSION TITLE
no running event loop after update to v 8.0

### DISCUSSION THREAD

**Discussion by mega15** [mitmproxy/mitmproxy#5274] [General]
I had a python script working on v7 but when I update the mitmproxy I face the following problem:

```
File "C:\dv\QA-BizagiProxy\bizproxy.py", line 86, in <module>
    m = DumpMaster(opts, with_termlog=True, with_dumper=False)
  File "C:\Python\Python39\lib\site-packages\mitmproxy\tools\dump.py", line 14, in __init__
    super().__init__(options)
  File "C:\Python\Python39\lib\site-packages\mitmproxy\master.py", line 31, in __init__
    self.event_loop = event_loop or asyncio.get_running_loop()
RuntimeError: no running event loop
```

my code is:

```
opts = options.Options(listen_host='127.0.0.1', listen_port=8080)
m = DumpMaster(opts, with_termlog=True, with_dumper=False)
addon = CaptureAddon(hosts, protocols)
m.addons.add(addon)

def loop_in_thread(loop, m):
asyncio.set_event_loop(loop)
m.run_loop(loop.run_forever)

loop = asyncio.get_event_loop()
t = threading.Thread( target=loop_in_thread, args=(loop,m) )
t.start()
```

Can anyone help me how to fix the problem?, thank you

**Comment by mhils** [comment]
See https://github.com/mitmproxy/mitmproxy/discussions/5255 (and indirectly https://stackoverflow.com/a/68661172/934719).

**Reply by mega15** [reply]
Thank you for the response. Following the idea of changing how to call the mitmproxy from my python script (after checking your answer), What can be the pattern code recommendation if I want the following code execution: 

1. Prepare my addons before calling th proxy (setting some variables and properties)
2. Call the mitmproxy asyncronous (not halt the main thread) 
3. Stop the proxy from my main code when some conditions were met 
4. Retrieve the result from the addons after the proxy was stopped

**Reply by mhils** [reply]
> 1. Prepare my addons before calling th proxy (setting some variables and properties)

I think the best way to do this is to define options for your addons: https://docs.mitmproxy.org/stable/addons-options/

> 2. Call the mitmproxy asyncronous (not halt the main thread)
> 3. Stop the proxy from my main code when some conditions were met

I would recommend running mitmdump in a separate thread using subprocess. We may make things easier here in the future, but this way will always work.

> 4. Retrieve the result from the addons after the proxy was stopped

Write to files (you can specify the path via addons).

I mean of course you can directly interface with DumpMaster/ConsoleMaster/WebMaster which makes some of these things easier, but then you need to expect that things may break in the future.

**Reply by BruceLee569** [reply]
> > 1. Prepare my addons before calling th proxy (setting some variables and properties)
> 
> I think the best way to do this is to define options for your addons: https://docs.mitmproxy.org/stable/addons-options/
> 
> > 2. Call the mitmproxy asyncronous (not halt the main thread)
> > 3. Stop the proxy from my main code when some conditions were met
> 
> I would recommend running mitmdump in a separate thread using subprocess. We may make things easier here in the future, but this way will always work.
> 
> > 4. Retrieve the result from the addons after the proxy was stopped
> 
> Write to files (you can specify the path via addons).
> 
> I mean of course you can directly interface with DumpMaster/ConsoleMaster/WebMaster which makes some of these things easier, but then you need to expect that things may break in the future.

I want to run mitmdump in the python code block so that I can control the input and output while the agent is running. 

After searching most of the answers on the Internet, it was found that the official did not provide this support.

In the end, I can only write the output from running mitmdump to the local file and read the file after the run is finished. 

Or run mitmdump through the subprocess module, and then read the file stream output of the subprocess.

These need to be communicated through io strings, and the implementation is not elegant enough. 

If mitmproxy has a better way to communicate with python namespace, please let me know.
